core/src/main/java/org/infinispan/transaction/xa/TransactionXaAdapter.java

```diff
@@ -57,23 +66,23 @@ public TransactionXaAdapter(XaTransactionTable txTable) {
     */
    @Override
    public int prepare(Xid externalXid) throws XAException {
-      return txTable.prepare(externalXid);
+      return joinAndRethrowXAException(CompletableFuture.supplyAsync(() -> txTable.prepare(externalXid), executor));
```

Comment:
Could use `CompletableFutures.COMPLETED_NULL.thenComposeAsync(...)` to avoid the double wrapping the result.
Maybe it's worth adding a static method in `CompletableFutures`.

Reply:
Ah, good one. Yeah, I thought there was a better way. Let me fix this. We may want it in CompletionStages though.